File: godor.cs
using System.Text;

var lines = File.ReadAllLines("melyseg.txt");
Console.WriteLine($"1. feladat\nA fájl adatainak száma: {lines.Length}\n");

Console.Write("2. feladat\nAdjon meg egy távolságértéket! ");
int chosenDistance = int.Parse(Console.ReadLine());
string chosenLine = lines[chosenDistance - 1];
Console.WriteLine($"Ezen a helyen a felszín {chosenLine} méter mélyen van.\n");

Console.WriteLine($"3. feladat\nAz érintetlen terület aránya {(double)lines.Where(x => x == "0").Count() / lines.Length * 100:0.00}%.\n");

int godorCount = 0;
bool prevGodor = false;
StringBuilder builder = new();
foreach (string line in lines)
{
    bool isGodor = line != "0";
    if (isGodor)
    {
        builder.Append(line + " ");
    }
    else if (prevGodor != isGodor)
    {
        builder.AppendLine();
        godorCount++;
    }
    prevGodor = isGodor;
}
File.WriteAllText("godrok.txt", builder.ToString());

Console.WriteLine($"5. feladat\nA gödrök száma: {godorCount}\n");

Console.WriteLine("6. feladat");
if (chosenLine == "0")
{
    Console.WriteLine("Az adott helyen nincs gödör.\n");
    return;
}

int first = 0;
int i = 0;
for (; i < chosenDistance - 1; i++)
{
    if (lines[i] == "0") first = i + 1;
}

int last = 0;
while (true)
{
    i++;
    if (lines[i] == "0")
    {
        last = i - 1;
        break;
    }
}

Console.WriteLine("a)");
Console.WriteLine($"A gödör kezdete: {first + 1} méter, a gödör vége: {last + 1} méter.");

Console.WriteLine("b)");
i = first;
bool isMonoton = true;
bool isDown = true;
int max = 0;
long sum = 0;
while (i <= last)
{
    int prev = int.Parse(lines[i - 1]);
    int curr = int.Parse(lines[i]);
    if (curr > max) max = curr;
    sum += curr;
    if (isDown)
    {
        if (curr < prev) isDown = false;
    }
    else
    {
        if (prev > curr) isMonoton = false;
    }
    i++;
}
Console.WriteLine(isMonoton
    ? "Folyamatosan mélyül."
    : "Nem mélyül folyamatosan.");

Console.WriteLine($"c)\nA legnagyobb mélysége {max} méter.");

Console.WriteLine($"d)\nA térfogata {sum * 10} m^3.");

Console.WriteLine($"e)\nA vízmennyiség {sum * 10 - (last - first + 1) * 10} m^3.");

File: godor.py
from distutils.file_util import write_file

class Melyseg:
    def __init__(self, index, melyseg) -> None:
        self.index = index
        self.melyseg = melyseg

file = open("melyseg.txt", "r")
lines = file.readlines()

melysegek = []
for line in lines:
    melysegek.append(int(line))

print("1. feladat:")
print(f"A fájl adatainak száma: {len(melysegek)}")

print("2. feladat")
index = int(input("Adjon meg egy távolságértéket! ")) - 1
print(f"Ezen a helyen a felszín {melysegek[index]} méter mélyen van.")

print("3. feladat")
erintetlen = sum(1 for x in melysegek if x == 0)
arany = erintetlen / len(melysegek)
print(f"Az érintetlen terület aránya: {arany * 100:.2f}%")

godrok: list[list[Melyseg]] = []
index_to_godor: dict[int, list[Melyseg]] = {}
i = 0
while True:
    while i < len(melysegek) and melysegek[i] == 0:
        i += 1
    if (i == len(melysegek)):
        break
    godrok.append([])
    while melysegek[i] != 0:
        godrok[-1].append(Melyseg(i, melysegek[i]))
        index_to_godor[i] = godrok[-1]
        i += 1
write_file("godrok.txt", (" ".join(str(melyseg.melyseg) for melyseg in godor) for godor in godrok))

print("5. feladat")
print(f"A gödrök száma: {len(godrok)}")

print("6. feladat")
if melysegek[index] == 0:
    print("Az adott helyen nincs gödör.")
else:
    print("a)")
    godor = index_to_godor[index]
    print(f"A gödör kezdete: {godor[0].index + 1} méter, a gödör vége: {godor[-1].index + 1} méter.")

    print("b)")
    legmelyebb = max(godor, key=lambda melyseg: melyseg.melyseg)
    legmelyebb_index = legmelyebb.index
    szmcs = sorted(godor[:legmelyebb_index], key=lambda melyseg: melyseg.melyseg) == godor[:legmelyebb_index]
    szmn = sorted(godor[legmelyebb_index:], key=lambda melyseg: melyseg.melyseg, reverse=True) == godor[legmelyebb_index:]
    if szmcs and szmn:
        print("Folyamatosan mélyül.")
    else:
        print("Nem mélyül folyamatosan.")

    print("c)")
    print(f"A legnagyobb mélysége {legmelyebb.melyseg} méter.")

    print("d)")
    print(f"A térfogata {sum(melyseg.melyseg for melyseg in godor) * 10} m^3.")

    print("e)")
    print(f"A vízmennyiség {(sum(melyseg.melyseg for melyseg in godor) - len(godor))* 10} m^3.")